Discuss inline functions and variables (#176)
hsf-training/cpluspluscourse: talk/basicconcepts/inline.tex

\subsection[inline]{Inline keyword}

\begin{frame}[fragile]
  \frametitlecpp[98]{Inline keyword}
  \begin{block}{Inline functions originally}
    \begin{itemize}
      \item can be applied to a function to tell the compiler to inline it
      \item just a hint, compiler can still choose to inline/not inline.
      \item effect: replace a function call by the function's content, like a macro
      \item avoids function call overhead, but may increase executable size if inlined in many places
      \item side effect: an inline function definition can be put into header files, the linker reduces the duplicated functions into one
    \end{itemize}

>   \end{block}

Author: Thx, I applied the suggested changes.

  \begin{block}{}
    \begin{cppcode*}{}
      inline int mult(int a, int b) {
        return a * b;
      }
    \end{cppcode*}
  \end{block}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[98]{Inline keyword}
  \begin{block}{Inline functions nowadays}
    \begin{itemize}

>       \item compilers can judge far better when to inline or not
>       \item side effect became primary purpose: to put functions into headers

Reviewer: ```suggestion
      \item compilers can judge far better when to inline or not
        \begin{itemize}
        \item thus primary purpose is gone
        \end{itemize}
      \item put functions into headers became main purpose
```

Author: Thx, I applied the suggested changes.

      \item many types of functions are marked \mintinline{cpp}{inline} by default:
      \begin{itemize}
        \item class member functions
        \item function templates
        \item \mintinline{cpp}{constexpr} functions
      \end{itemize}
    \end{itemize}
  \end{block}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[17]{Inline keyword}
  \begin{block}{Inline variables}
    \begin{itemize}
      \item we can also mark a global (or \mintinline{cpp}{static} member) variable as inline
      \item same side effect, linker merges all occurrences into one

>       \item useful for defining global variables/constants in headers

Reviewer: ```suggestion
      \item allows to define global variables/constants in headers
```

Author: Yeah, let's not endorse that too much :D
